src/socket/tcp/congestion_controller.rs

```diff
@@ -0,0 +1,90 @@
+use crate::time::Instant;
+
+use super::RttEstimator;
+
+pub(super) mod cubic;
+pub(super) mod reno;
+
+pub(super) trait CongestionController: Default {
+    /// Returns the number of bytes that can be sent.
+    fn window(&self) -> usize;
+
+    /// Set the remote window size.
+    #[allow(unused_variables)]
+    fn set_remote_window(&mut self, remote_window: usize) {}
+
+    #[allow(unused_variables)]
+    fn on_ack(&mut self, now: Instant, len: usize, rtt: &RttEstimator) {}
+
+    #[allow(unused_variables)]
+    fn on_retransmit(&mut self, now: Instant) {}
+
+    #[allow(unused_variables)]
+    fn on_duplicate_ack(&mut self, now: Instant) {}
+
+    #[allow(unused_variables)]
+    fn pre_transmit(&mut self, now: Instant) {}
+
+    #[allow(unused_variables)]
+    fn post_transmit(&mut self, now: Instant, len: usize) {}
+
+    /// Set the maximum segment size.
+    #[allow(unused_variables)]
+    fn set_mss(&mut self, mss: usize) {}
+}
+
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub(super) enum GenericController {
+    None,
+    Cubic(cubic::Cubic),
+    Reno(reno::Reno),
+}
+
+impl Default for GenericController {
+    fn default() -> Self {
+        GenericController::None
+    }
+}
+
+impl CongestionController for GenericController {
```

Comment:
Can this be implemented as something like the below rather than with this indirect impl?

```
impl GenericController {
    fn inner_mut(&mut self) -> &mut dyn Trait {
    }
}
```

Reply:
I tried it, but Rust compiler reject it.

Reply:
I tried it just now in the playground and it seems to work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ee36a332e85a0b3d92f1a437f1b949da

Reply:
If you do it like this, I think that in the case where the enum has only one element, the call might get devirtualized, and you'll get the benefits of the generic parameter without actually using one.

Reply:
Because `Default` trait requires `Self: Sized`, `inner_mut()` has been rejected.
I will try to avoid `Default`.

Reply:
Removing `Default` from `CongestionController`'s types, and `inner_mut()` works fine.
Thank you!